File: exercises/cookbook/flags.cpp
// Appendix F, Recipes 31 and 32 - a feature flag read once at startup and
// kept as a member, and a [Flags] enum spelled as an enum class with its
// operators.
//
// Features and Processor (Recipe 31), and Channel with its three operators
// (Recipe 32), are quoted VERBATIM in book/F-rosetta-cookbook.md: editing
// one means editing the appendix in the same commit (the testlab
// discipline). main() is scaffolding - it sets the environment, reads the
// flags once, asserts the branch, the read-once, the junk-keeps-the-default
// parse and the bit arithmetic; the refusals the recipes name stay comments.
#include <cassert>
#include <charconv>
#include <cstdint>
#include <cstdlib>
#include <string_view>

// Recipe 31 - IConfiguration read at startup, once
struct Features {
    bool audit = false;                  // the defaults ARE the off state
    bool fast_path = false;
    int  batch_size = 64;

    // One source among several - Recipe 26's JSON file, the host's
    // preferences API, a command line. Whatever the source, it is read HERE,
    // once, and never again.
    static Features from_environment() {
        Features f;
        if (const char* v = std::getenv("MYPLUGIN_AUDIT"))      f.audit = std::string_view(v) == "1";
        if (const char* v = std::getenv("MYPLUGIN_FAST_PATH"))  f.fast_path = std::string_view(v) == "1";
        if (const char* v = std::getenv("MYPLUGIN_BATCH_SIZE")) {
            const std::string_view s(v);
            std::from_chars(s.data(), s.data() + s.size(), f.batch_size);   // junk: batch_size stays 64 (Recipe 19)
        }
        return f;
    }
};

class Processor {
public:
    explicit Processor(Features features) : features_(features) {}   // read once, kept as a member

    int process(int sample) const {
        if (features_.fast_path) {       // a branch: free, even on the deadline path
            return sample;
        }
        return sample * 2;
    }

private:
    Features features_;
};

// Recipe 32 - [Flags] enum Channel, and HasFlag
enum class Channel : std::uint8_t { None = 0, Left = 1, Right = 2, Sub = 4 };   // [Flags] enum Channel

constexpr Channel operator|(Channel a, Channel b) {
    return static_cast<Channel>(static_cast<std::uint8_t>(a) | static_cast<std::uint8_t>(b));
}
constexpr Channel operator&(Channel a, Channel b) {
    return static_cast<Channel>(static_cast<std::uint8_t>(a) & static_cast<std::uint8_t>(b));
}
constexpr bool has(Channel set, Channel flag) { return (set & flag) == flag; }   // set.HasFlag(flag)

namespace {
    void set_env(const char* name, const char* value) {
#if defined(_WIN32)
        _putenv_s(name, value);
#else
        setenv(name, value, 1);
#endif
    }
}

int main() {
    // Nothing set: the defaults are the off state.
    const Features off = Features::from_environment();
    assert(!off.audit && !off.fast_path && off.batch_size == 64);
    assert(Processor(off).process(21) == 42);

    // Set, read once, kept: the environment can change afterwards and the
    // Processor does not care - that is the whole point of reading once. (A
    // Processor that re-read the environment in process() fails the last
    // assert here; it was run.) The broken shape the recipe's trap names -
    // the read at the point of use on the deadline path - would pass every
    // assertion in this file and stays book-only.
    set_env("MYPLUGIN_FAST_PATH", "1");
    set_env("MYPLUGIN_BATCH_SIZE", "128");
    const Features on = Features::from_environment();
    assert(on.fast_path && on.batch_size == 128 && !on.audit);
    const Processor fast(on);
    set_env("MYPLUGIN_FAST_PATH", "0");
    assert(fast.process(21) == 21);

    // Junk in the environment keeps the default - from_chars leaves the
    // target untouched on failure, where atoi would have written 0 over 64.
    set_env("MYPLUGIN_BATCH_SIZE", "abc");
    assert(Features::from_environment().batch_size == 64);

    // The [Flags] enum: combine, test, and the type survives the arithmetic.
    constexpr Channel stereo = Channel::Left | Channel::Right;
    static_assert(has(stereo, Channel::Left));
    static_assert(has(stereo, Channel::Right));
    static_assert(!has(stereo, Channel::Sub));
    static_assert(has(stereo | Channel::Sub, Channel::Sub));
    static_assert(!has(Channel::None, Channel::Left));
    static_assert((stereo & Channel::Sub) == Channel::None);
    static_assert(has(Channel::Left, Channel::None));   // like HasFlag(0): always true - test bits, not None
    // The trap the recipe names: an overlap test reads as HasFlag and is
    // wrong for a combined flag - a set holding only Left "has" stereo.
    static_assert((Channel::Left & stereo) != Channel::None);   // the wrong test says yes
    static_assert(!has(Channel::Left, stereo));                 // has() says no
    // The refusal, as a comment because a refusal cannot be compiled:
    // without the operators above, `Channel::Left | Channel::Right` is
    // "invalid operands to binary expression" (clang; GCC says "no match
    // for 'operator|'") - an enum class does not decay to int, which is the
    // feature. A plain `enum` would compile it and hand back an int, and
    // the type would be gone.
    return 0;
}

File: exercises/cookbook/rust/src/flags.rs
//! Appendix F, Recipes 31 and 32 - a feature flag read once, and flags
//! combined as a type. Included by book/F-rosetta-cookbook.md between the
//! recipe-N markers: edit here and the page follows. The tests are
//! scaffolding and appear in no listing.

// --8<-- [start:recipe-31]
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Features {
    pub audit: bool,         // the defaults ARE the off state
    pub fast_path: bool,
    pub batch_size: usize,
}

impl Default for Features {
    fn default() -> Self {
        Self { audit: false, fast_path: false, batch_size: 64 }
    }
}

impl Features {
    // One source among several. Whatever the source, it is read HERE, once, and never again.
    pub fn from_environment() -> Self {
        let flag = |name: &str| std::env::var(name).map(|v| v == "1").unwrap_or(false);
        let mut f = Self { audit: flag("MYPLUGIN_AUDIT"), fast_path: flag("MYPLUGIN_FAST_PATH"), ..Self::default() };
        if let Ok(size) = std::env::var("MYPLUGIN_BATCH_SIZE") {
            f.batch_size = size.parse().unwrap_or(f.batch_size);    // junk: batch_size stays 64 (Recipe 19)
        }
        f
    }
}

pub struct Processor {
    features: Features,    // read once, kept as a field
}

impl Processor {
    pub fn new(features: Features) -> Self {
        Self { features }
    }

    pub fn process(&self, sample: i32) -> i32 {
        if self.features.fast_path { sample } else { sample * 2 }    // a branch: free, even on the deadline path
    }
}
// --8<-- [end:recipe-31]

// --8<-- [start:recipe-32]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Channel(u8);    // [Flags] enum Channel: a newtype over the bits, not an enum

impl Channel {
    pub const NONE: Channel = Channel(0);
    pub const LEFT: Channel = Channel(1);
    pub const RIGHT: Channel = Channel(2);
    pub const SUB: Channel = Channel(4);

    pub const fn has(self, flag: Channel) -> bool {
        self.0 & flag.0 == flag.0    // set.HasFlag(flag)
    }
}

impl std::ops::BitOr for Channel {
    type Output = Channel;
    fn bitor(self, rhs: Channel) -> Channel {
        Channel(self.0 | rhs.0)
    }
}

impl std::ops::BitAnd for Channel {
    type Output = Channel;
    fn bitand(self, rhs: Channel) -> Channel {
        Channel(self.0 & rhs.0)
    }
}
// --8<-- [end:recipe-32]

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flags_read_once_and_junk_keeps_the_default() {
        std::env::set_var("MYPLUGIN_FAST_PATH", "1");
        std::env::set_var("MYPLUGIN_BATCH_SIZE", "lots");
        let features = Features::from_environment();
        assert_eq!(features, Features { audit: false, fast_path: true, batch_size: 64 });
        let processor = Processor::new(features);
        std::env::set_var("MYPLUGIN_FAST_PATH", "0");    // changed after the read: no effect
        assert_eq!(processor.process(21), 21);
        assert_eq!(Processor::new(Features::default()).process(21), 42);
    }

    #[test]
    fn bits_combine_and_test() {
        let stereo = Channel::LEFT | Channel::RIGHT;
        assert!(stereo.has(Channel::LEFT) && stereo.has(Channel::RIGHT) && !stereo.has(Channel::SUB));
        assert_eq!(stereo & Channel::SUB, Channel::NONE);
        assert!(Channel::NONE.has(Channel::NONE));    // HasFlag(None) is true in C# too
    }
}
